docs/architecture/adr-082-socket-protocol.md

```diff
@@ -0,0 +1,237 @@
+# ADR 082: The Future of the Socket Protocol
+
+## Changelog
+
+- 19-May-2022: Initial draft (@creachadair)
+
+## Status
+
+(in review)
+
+## Context
+
+The [Application Blockchain Interface (ABCI)][abci] is a client-server protocol
+used by the Tendermint consensus engine to communicate with the application on
+whose behalf it performs state replication. There are currently three transport
+options available for ABCI applications:
+
+1. **In-process**: Applications written in Go can be linked directly into the
+   same binary as the consensus node. Such applications use a "local" ABCI
+   connection, which exposes application methods to the node as direct function
+   calls.
```

Comment:
I'm thinking whether this option could be provided to other languages... via C FFI + dynamic linking (perhaps https://github.com/petermattis/fastcgo if cgo is a concern).

and perhaps an extension to protoc can be developed (e.g. based on https://github.com/protobuf-c/protobuf-c) to generate the C FFI shim for each language (such as https://cython.readthedocs.io/en/latest/src/tutorial/embedding.html )

Reply:
There's also inproc transport for gGRPC: https://github.com/grpc/grpc/tree/master/src/core/ext/transport/inproc
but it seems it's not ideal: https://groups.google.com/g/grpc-io/c/jAvNYl-_i30

Reply:
I've thought about this approach, but I don't think it's the right one.

An FFI interface is possible for a lot of languages, but I don't think it's the right choice for a general solution. FFI support varies widely in coverage and quality across languages (and the points of friction can be tricky to work around). Moreover, it's much harder to add after-the-fact for an application developer. On the Go side it's not that difficult, though in practice the C shims can get complicated if there's a lot of variability in the runtime environment on the other end.

From an engineering perspective I think the in-process option was a premature optimization, and if we were starting over today I would probably not provide it even for Go applications. The overhead of local IPC is entirely dominated by other aspects of consensus—both on the application side, e.g., executing transactions, writing to persistent storage—and the network costs of gossiping transactions, votes, and evidence.

I think enforcing process isolation could also improve both operational issues (responding consistently to application failures, for instance) and correctness concerns (an in-process application could easily corrupt consensus state, either by accident or intent).

We probably can't get rid of the existing Go in-process support, but I don't think we should try to extend it. If we hope to have one answer for non-Go applications I think we're better off picking an IPC-based solution (whether that's gRPC or an extension of our custom socket protocol or something else).